lms/views/api/error.py

```diff
@@ -12,52 +13,125 @@
+       opens "auth_url".
 
-@forbidden_view_config(path_info="/api/*", renderer="json")
-def forbidden(request):
-    request.response.status_int = 403
-    return {"message": _("You're not authorized to view this page")}
+       If no "message" is present the frontend will show a standard
+       authorization dialog (not an error dialog) and the button that opens
+       "auth_url" will be labelled [Authorize].
 
+    3. "details": Optional further error details to show to the user after
+       "message", for debugging and support.
 
-@notfound_view_config(path_info="/api/*", renderer="json")
-def notfound(request):
-    request.response.status_int = 404
-    return {"message": _("Endpoint not found")}
+    The HTTP status codes of the API error responses are:
 
+    * 403 Forbidden is used if the frontend's authentication to the backend's
+      API failed
 
-@exception_view_config(path_info="/api/*", context=Exception, renderer="json")
-def api_error(request):
-    """Fallback error handler for frontend API requests."""
-    request.response.status_int = 500
+    * 400 Bad Request is deliberately abused whenever the backend's
+      server-to-server request to a third-party API (such as the Canvas API or
+      the Blackboard API) fails or can't be made for any reason.
 
-    # Exception details are not reported here to avoid leaking internal information.
-    return {
-        "message": "A problem occurred while handling this request. Hypothesis has been notified."
-    }
+      For example if we can't authenticate to the third-party API because we
+      don't have an access token for this user yet; if we get an authentication
+      error from the third-party API because the access token has expired (and
+      can't be refreshed); if the third-party API returns an invalid,
+      unexpected or unsuccessful response; are all 400s.
+
+      The frontend uses the JSON bodies of these 400 Bad Request responses to
+      distinguish between them and decide what to do.
+
+      Since 403 is used to mean that the frontend's authentication *to the
+      backend* failed, it can't also be used to indicate failure of the
+      backend's server-to-server request to authenticate with the third-party
+      API. Hence why 400 is used for this.
+
+      You might think that some sort of gateway error (e.g. 502 Bad Gateway) is
+      more semantically correct than abusing 400 here. But Cloudflare replaces
+      502 JSON responses from our app with its own Cloudflare error pages. So
+      we can't use 5xx statuses in production.
+
+    * 422 is used for validation errors (if the frontend sent invalid params to
+      a backend API) (in practice this should never happen)
+
+    * 404 Not Found is used for API endpoints that don't exist
+      (in practice this should never happen)
+
+    * 500 Server Error is used for unexpected exceptions (our code crashed)
+      (in practice this should never happen)
```

Comment:
Took the opportunity to add a long docstring explaining how the API error responses work. This has caused confusion before so it seemed worth documenting.